### PR TITLE
Keep Selected Area on Refresh + Warning for Long URL Length (Geospatial)

### PR DESCRIPTION
**Pull request by zandre-eng**
## Product Description

When a saved polygon is selected for filtering it will now continue to be selected after a page refresh. Previously, upon refreshing the page the selected saved polygon would be added to the map, however the selection dropdown would show no saved polygon selected.

Additionally, if a user adds too many polygons to the map and causes the URL length to be too long, a warning will be displayed to the user:
url-length-warning

## Technical Summary

Link to tickets here:
- [BadRequestError](https://dimagi.atlassian.net/browse/SC-3272)
- [Filtered Saved Area is not displayed after refresh](https://dimagi.atlassian.net/browse/SC-3271)

This PR addresses an issue on the Geospatial Case Grouping page where users could potentially have a URL length that too long (>4500 characters). This is because when polygons for filtering are added to the map, their JSON gets added into the URL to be sent to the back-end. If a user adds too many polygons to the map then they will receive a `BadRequestError`. This has been addressed by not adding additional polygon JSON to the URL if it will result in the URL being too long. Instead, a relevant warning is given to the user.

Additionally, since every map polygon's JSON is stored in the URL in a single param, there was no easy way to identify and re-select the selected saved polygon on a page refresh. This PR addresses the issue by saving only the selected polygon ID in the URL (not its entire JSON). The selected polygon JSON gets retrieved in the back-end by looking for the correct `GeoPolygon` instance using the given selected polygon ID.

## Feature Flag

`GEOSPATIAL`

### Safety story

- Local testing
- Unit tests

### Automated test coverage

Automated tests exist for `CaseGroupingReport` class, including the new `features_from_request` class property that was added in this PR.

### QA Plan

No QA planned.

### Rollback instructions

- [X] This PR can be reverted after deploy with no further considerations

### Labels & Review
- [X] Risk label is set correctly
- [X] The set of people pinged as reviewers is appropriate for the level of risk of the change

### CODE REVIEW

File: corehq/apps/geospatial/static/geospatial/js/models.js [hunk 1]
```diff
@@ -657,7 +673,7 @@ hqDefine('geospatial/js/models', [
             createActivePolygonLayer(polygonObj);
 
             self.activeSavedPolygon = polygonObj;
-            self.addPolygonsToFilterList(polygonObj.geoJson.features);
```
Comment (mkangia) [inline corehq/apps/geospatial/static/geospatial/js/models.js:647]: is this method not needed anymore?

File: corehq/apps/geospatial/reports.py [hunk 2]
```diff
@@ -277,6 +278,22 @@ def _aggregate_query(self, query):
...
+    @property
+    def features_from_request(self):
+        features = {}
+        features_json = self.request.GET.get('features')
+        if features_json:
+            try:
+                features |= json.loads(features_json)
+            except json.JSONDecodeError:
+                raise ValueError(f'{features_json!r} parameter is not valid JSON')
+        selected_feature_id = self.request.GET.get('selected_feature_id')
```
Comment (mkangia) [inline corehq/apps/geospatial/reports.py:290]: nit: how about extracting a method `_expand_geo_polygon`?
I am not sure if editing the feature is the right thing to do here. If someone refers to `features_from_request`, they would assume its the one that is exactly what was set in the request.

File: corehq/apps/geospatial/templates/case_grouping_map.html [hunk 3]
```diff
@@ -53,6 +53,12 @@
       to apply the polygon filtering changes.
     {% endblocktrans %}
   </div>
+  <div class="alert alert-warning" data-bind="visible: hasUrlError">
+    {% blocktrans %}
+      There are too many polygons on the map. Please remove some before refreshing the page, otherwise the
+      newly added polygons will be ignored when filtering.
```
[the remark this answers is not in the document]
Comment (zandre-eng) [reply corehq/apps/geospatial/templates/case_grouping_map.html:59]: That is correct. If the new polygon causes the URL length to exceed max then it won't be added to the URL, however we'll still keep track of the new polygon in `self.polygons`. This means that if a user deletes a polygon then the newly added polygon should get added to the URL. In saying that, I wonder if we should keep track of the new polygon or just ignore it completely?

File: corehq/apps/geospatial/static/geospatial/js/models.js [hunk 4]
```diff
@@ -14,6 +14,7 @@ hqDefine('geospatial/js/models', [
 ) {
     const DOWNPLAY_OPACITY = 0.2;
     const FEATURE_QUERY_PARAM = 'features';
+    const MAX_URL_LENGTH = 4500;
```
[the remark this answers is not in the document]
Comment (zandre-eng) [reply corehq/apps/geospatial/static/geospatial/js/models.js:18]: This comes from a configuration setting in gunicorn [here](https://github.com/dimagi/commcare-hq/blob/e673e475f7eb20c1495a02a3ca56d805022f7616/deployment/gunicorn/gunicorn_conf.py#L13) where we specify the max client URL length.

File: corehq/apps/geospatial/static/geospatial/js/models.js (hunk 1 above)
Comment (zandre-eng) [reply corehq/apps/geospatial/static/geospatial/js/models.js:647]: This method still gets used by the "Case Grouping" page [here](https://github.com/dimagi/commcare-hq/blob/e673e475f7eb20c1495a02a3ca56d805022f7616/corehq/apps/geospatial/static/geospatial/js/case_grouping_map.js#L543) and [here](https://github.com/dimagi/commcare-hq/blob/e673e475f7eb20c1495a02a3ca56d805022f7616/corehq/apps/geospatial/static/geospatial/js/case_grouping_map.js#L549). It is being used to add polygons to the URL when a user draws/moves/deletes a new polygon on the map.

File: corehq/apps/geospatial/reports.py (hunk 2 above)
Comment (zandre-eng) [reply corehq/apps/geospatial/reports.py:290]: > how about extracting a method _expand_geo_polygon?

Sure, I could alternatively extract this to a `saved_polygon_from_request` function. That way we can make it more explicit which features are coming from the request JSON and which are being retrieved from the back-end. What are your thoughts?

> I am not sure if editing the feature is the right thing to do here.

Just noting that we're not explicitly editing the saved features here but rather just parsing them into a dict format with randomly generated IDs (saved features don't have an ID). This is being done to keep it in the same format as the dict of features coming from the front-end.

File: corehq/apps/geospatial/templates/case_grouping_map.html (hunk 3 above)
[the remark this answers is not in the document]
Comment (mkangia) [reply corehq/apps/geospatial/templates/case_grouping_map.html:59]: Hey @zandre-eng 

I am a bit confused by that.

My original feedback was just regarding the message shown to the user. I was thinking the message should be updated to
"otherwise the new added polygons will be lost".

About the cases you mentioned, I am unable to put that together.

File: corehq/apps/geospatial/reports.py (hunk 2 above)
Comment (mkangia) [reply corehq/apps/geospatial/reports.py:290]: `saved_polygon_from_request` sounds good to me.

> Just noting that we're not explicitly editing the saved features 

Right. I just meant to say the value of features coming in the request.

File: corehq/apps/geospatial/templates/case_grouping_map.html (hunk 3 above)
[the remark this answers is not in the document]
Comment (zandre-eng) [reply corehq/apps/geospatial/templates/case_grouping_map.html:59]: > My original feedback was just regarding the message shown to the user.

Ah. apologies for the misunderstanding, I thought you wanted to know more about how we handle the logic for when there are too many polygons. The suggested wording sounds good to me.

> About the cases you mentioned, I am unable to put that together.

Please do let me know if you'd like me to clear up any confusion regarding my previous comment and how we handle too many polygons in the code.

File: corehq/apps/geospatial/reports.py (hunk 2 above)
Comment (zandre-eng) [reply corehq/apps/geospatial/reports.py:295]: [c9476db](https://github.com/dimagi/commcare-hq/pull/34511/commits/c9476db4fa2369002a2ccc78be1eee8e97c18822)

File: corehq/apps/geospatial/templates/case_grouping_map.html (hunk 3 above)
[the remark this answers is not in the document]
Comment (zandre-eng) [reply corehq/apps/geospatial/templates/case_grouping_map.html]: [ca9610b](https://github.com/dimagi/commcare-hq/pull/34511/commits/ca9610b06d15bb13768ddfcecdac63ca7ff51a30)

File: corehq/apps/geospatial/reports.py [hunk 5]
```diff
@@ -277,6 +278,31 @@ def _aggregate_query(self, query):
             precision = find_precision(query, case_property)
         return apply_geohash_agg(query, case_property, precision)
 
+    @property
+    def features_from_request(self):
+        features = {}
+        features_json = self.request.GET.get('features')
+        if features_json:
+            try:
+                features |= json.loads(features_json)
```
Comment (mkangia) [inline corehq/apps/geospatial/reports.py:287]: nit: just curious about the usage of `|=` here. Just the usual `=` won't do?

File: corehq/apps/geospatial/reports.py (hunk 2 above)
Comment (mkangia) [reply corehq/apps/geospatial/reports.py:295]: Thanks @zandre-eng 

nit: Should the name be `selected_polygon_id` of the variable and of the request param? or it could be anything other than a polygon as well?

File: corehq/apps/geospatial/reports.py [hunk 6]
```diff
@@ -277,6 +278,31 @@ def _aggregate_query(self, query):
...
+        return features
+
+    @property
+    def saved_polygon_from_request(self):
+        features = {}
+        selected_feature_id = self.request.GET.get('selected_feature_id')
+        if selected_feature_id:
+            selected_polygon = GeoPolygon.objects.get(id=selected_feature_id)
+            geo_json_features = selected_polygon.geo_json['features']
+            features = {uuid.uuid4().hex: feature for feature in geo_json_features}
```
Comment (mkangia) [inline corehq/apps/geospatial/reports.py:299]: nit: how is the hex used?

File: corehq/apps/geospatial/static/geospatial/js/models.js [hunk 7]
```diff
@@ -591,12 +628,12 @@ hqDefine('geospatial/js/models', [
...
         self.clearActivePolygon = function () {
             if (self.activeSavedPolygon) {
-                // self.selectedSavedPolygonId('');
-                self.removePolygonsFromFilterList(self.activeSavedPolygon.geoJson.features);
+                self.selectedSavedPolygonId('');
                 removeActivePolygonLayer();
                 self.activeSavedPolygon = null;
                 self.btnSaveDisabled(false);
                 self.btnExportDisabled(true);
+                updateSelectedSavedPolygonParam();
```
Comment (mkangia) [inline corehq/apps/geospatial/static/geospatial/js/models.js:636]: is it possible to instead subscribe to change to `selectedSavedPolygonId` to call this function?

**Review by mkangia — APPROVED** [review]
Thanks for the follow up @zandre-eng 

Just some non blocking feedback

File: corehq/apps/geospatial/reports.py (hunk 5 above)
Comment (zandre-eng) [reply corehq/apps/geospatial/reports.py:287]: Ah, you are correct in that we can simply use `=`. This was old code from when we were still adding both the features and selected polygon to the same dict. I'll fix this.

File: corehq/apps/geospatial/reports.py (hunk 6 above)
Comment (zandre-eng) [reply corehq/apps/geospatial/reports.py:299]: I used the hex format to keep the ID format consistent with the dict of features coming from `features_from_request` which also use a hex UUID for its keys. We don't use the IDs/dict keys directly in the back-end code, but good to keep the same for consistency's sake.

**Comment by zandre-eng** [comment]
> Should the name be selected_polygon_id of the variable and of the request param? or it could be anything other than a polygon as well?

@mkangia I used `selected_feature_id` here to keep the naming consistent with how we use it throughout the code (e.g. [here](https://github.com/dimagi/commcare-hq/blob/95b2b98fce61b4351bfa72cd7e44920a8f180297/corehq/apps/geospatial/reports.py#L304) and [here](https://github.com/dimagi/commcare-hq/blob/95b2b98fce61b4351bfa72cd7e44920a8f180297/corehq/apps/geospatial/reports.py#L229)). This is also in line with the naming convention that MapBox uses. For our use case it should only be a polygon (as this is what is supported in ES).

File: corehq/apps/geospatial/static/geospatial/js/models.js (hunk 7 above)
Comment (zandre-eng) [reply corehq/apps/geospatial/static/geospatial/js/models.js]: Definitely. I see we are already calling `updateSelectedSavedPolygonParam()` in the subscribe function for `selectedSavedPolygonId` so can remove it from here ([7f5da66](https://github.com/dimagi/commcare-hq/pull/34511/commits/7f5da662bf8de6c0c5901313594e2291ff0a9d2d)).

File: corehq/apps/geospatial/reports.py (hunk 5 above)
Comment (zandre-eng) [reply corehq/apps/geospatial/reports.py]: [253501a](https://github.com/dimagi/commcare-hq/pull/34511/commits/253501a18f7586e154cb143acfbc60891af3b747)

File: corehq/apps/geospatial/static/geospatial/js/models.js [hunk 8]
```diff
@@ -628,12 +628,10 @@ hqDefine('geospatial/js/models', [
 
         self.clearActivePolygon = function () {
             if (self.activeSavedPolygon) {
-                self.selectedSavedPolygonId('');
```
Comment (mkangia) [inline corehq/apps/geospatial/static/geospatial/js/models.js]: why was this removed?

Comment (zandre-eng) [reply corehq/apps/geospatial/static/geospatial/js/models.js]: I realised this actually introduced a bug where selecting a saved polygon resulted in the dropdown being cleared. This is because we call `clearActivePolygon()` before adding the new saved polygon to the map.

Comment (mkangia) [reply corehq/apps/geospatial/static/geospatial/js/models.js]: okay.

**Review by mkangia — APPROVED** [review]
